src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java

```diff
@@ -0,0 +1,367 @@
+		runSequentially(invocation, invocationContext, extensionContext);
+	}
+
+	@Override
+	public void interceptAfterEachMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		runSequentially(invocation, invocationContext, extensionContext);
+	}
+
+	private <T> T runSequentially(Invocation<T> invocation,
+			ReflectiveInvocationContext<? extends Executable> invocationContext, ExtensionContext extensionContext)
+			throws Throwable {
+		ExtensionContext.Store store = extensionContext.getRoot().getStore(NAMESPACE);
+		List<ReentrantLock> locks = findSharedOnExecutable(invocationContext.getExecutable())
+				// Sort by @Shared's name to prevent deadlocks when locking later.
+				// This is a well-known solution to the "dining philosophers problem".
+				// See ResourcesParallelismTests.ThrowIfTestsRunInParallelTestCases for more info.
+				.sorted(comparing(Shared::name))
+				.map(shared -> findLockForShared(shared, store))
+				.collect(toList());
+		return invokeWithLocks(invocation, locks);
+	}
+
+	private void runDynamicTestSequentially(Invocation<Void> invocation, ExtensionContext extensionContext)
+			throws Throwable {
+		ExtensionContext.Store store = extensionContext.getRoot().getStore(NAMESPACE);
+		List<ReentrantLock> locks = findSharedOnExecutable(testFactoryMethod(extensionContext))
+				// see runSequentially
+				.sorted(comparing(Shared::name))
+				.map(shared -> findLockForShared(shared, store))
+				.collect(toList());
+		invokeWithLocks(invocation, locks);
+	}
+
+	private Method testFactoryMethod(ExtensionContext extensionContext) {
+		return extensionContext
+				.getParent()
+				.orElseThrow(() -> new IllegalStateException(
+					"The parent extension context of a DynamicTest was not a @TestFactory-annotated test method"))
+				.getRequiredTestMethod();
+	}
+
+	private Stream<Shared> findSharedOnExecutable(Executable executable) {
+		return Arrays
+				.stream(executable.getParameters())
+				.map(parameter -> AnnotationSupport.findAnnotation(parameter, Shared.class))
+				.filter(Optional::isPresent)
+				.map(Optional::get);
+	}
+
+	private ReentrantLock findLockForShared(Shared shared, ExtensionContext.Store store) {
+		return Optional
+				.ofNullable(store.get(resourceKey(shared), ResourceWithLock.class))
+				.orElseThrow(
+					() -> new IllegalStateException("There should be a shared resource for the name " + shared.name()))
+				.lock();
+	}
+
+	private static <T> T invokeWithLocks(Invocation<T> invocation, List<ReentrantLock> locks) throws Throwable {
+		locks.forEach(ReentrantLock::lock);
+		try {
+			return invocation.proceed();
+		}
+		finally {
+			// for dining philosophers, "[t]he order in which each philosopher puts down the forks does not matter"
+			// (quote from Wikipedia)
+			locks.forEach(ReentrantLock::unlock);
```

Comment:
I changed this from reverse unlocking because I'm pretty sure we don't need that.

Reply:
Oh excellent, thanks for making the fix!